### ISSUE TITLE
FSDP2 error loading `PreTrainedModel`'s `torch.Tensor` state_dict into `DTensor`

### ISSUE THREAD

**Issue by ringohoffman** [pytorch/pytorch#136336]
### 🐛 Describe the bug

I'm trying to follow the instructions to efficiently load Hugging Face models from [`torchtitan`'s docs for FSDP1 -> FSDP2: Meta-Device Initialization](https://github.com/pytorch/torchtitan/blob/d2a4904f58accc683c17c66a360026cb3c8109af/docs/fsdp.md#meta-device-initialization). When I get to the last step of actually loading the model's weights, I get an error about trying to copy `torch.Tensor` into the FSDP2 `DTensor`s.

```python
"""To run:

pip install torch==2.4.1 accelerate==0.34.0 transformers==4.44.2

OMP_NUM_THREADS=2 \
TOKENIZERS_PARALLELISM=false \
CUDA_VISIBLE_DEVICES=6,7 \
torchrun \
    --rdzv-backend=c10d \
    --rdzv-endpoint=localhost:0 \
    --nnodes=1 \
    --nproc-per-node=2 \
    fsdp_generate.py
"""
import accelerate
import torch
import torch.distributed
import torch.distributed._composable.fsdp
import transformers
import transformers.models.gpt_neo.modeling_gpt_neo


def main() -> None:
    torch.cuda.set_device(torch.device(torch.distributed.get_rank()))

    # creating a checkpoint we can load with load_checkpoint_and_dispatch
    transformers.AutoModelForCausalLM.from_pretrained(pretrained_model_name_or_path := "EleutherAI/gpt-neo-125m").save_pretrained(checkpoint := "checkpoint")

    # following the steps from: https://github.com/pytorch/torchtitan/blob/d2a4904f58accc683c17c66a360026cb3c8109af/docs/fsdp.md
    with torch.device("meta"):
        config = transformers.AutoConfig.from_pretrained(pretrained_model_name_or_path, torch_dtype=torch.bfloat16)
        model = transformers.models.gpt_neo.modeling_gpt_neo.GPTNeoForCausalLM(config)

    for submodule in model.modules():
        if isinstance(submodule, transformers.models.gpt_neo.modeling_gpt_neo.GPTNeoBlock):
            torch.distributed._composable.fsdp.fully_shard(submodule)
    torch.distributed._composable.fsdp.fully_shard(model)

    model.to_empty(device="cuda")

    # this boils down to load_state_dict
    accelerate.load_checkpoint_and_dispatch(model, checkpoint, dtype=torch.bfloat16)


if __name__ == "__main__":
    torch.distributed.init_process_group(world_size=2)
    try:
        main()
    finally:
        torch.distributed.destroy_process_group()

```

```
...
[rank1]:    While copying the parameter named "transformer.ln_f.weight", whose dimensions in the model are torch.Size([768]) and whose dimensions in the checkpoint are torch.Size([768]), an exception occurred : ('aten.copy_.default: got mixed torch.Tensor and DTensor, need to convert all torch.Tensor to DTensor before calling distributed operators!',).
[rank1]:    While copying the parameter named "transformer.ln_f.bias", whose dimensions in the model are torch.Size([768]) and whose dimensions in the checkpoint are torch.Size([768]), an exception occurred : ('aten.copy_.default: got mixed torch.Tensor and DTensor, need to convert all torch.Tensor to DTensor before calling distributed operators!',).
```

Coming from:

https://github.com/pytorch/pytorch/blob/172ecf78b75f55948f9102c85e72314d1f8f1a22/torch/distributed/tensor/_dispatch.py#L458-L473

`accelerate.load_checkpoint_and_dispatch` is just calling `load_state_dict` under the hood. How are users expected to initialize from pretrained weights in this scenario?

@awgu @wconstab 

### Versions

Collecting environment information...
PyTorch version: 2.4.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-119-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3
GPU 2: NVIDIA H100 80GB HBM3
GPU 3: NVIDIA H100 80GB HBM3
GPU 4: NVIDIA H100 80GB HBM3
GPU 5: NVIDIA H100 80GB HBM3
GPU 6: NVIDIA H100 80GB HBM3
GPU 7: NVIDIA H100 80GB HBM3

Nvidia driver version: 535.183.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               224
On-line CPU(s) list:                  0-223
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8480+
CPU family:                           6
Model:                                143
Thread(s) per core:                   2
Core(s) per socket:                   56
Socket(s):                            2
Stepping:                             8
Frequency boost:                      enabled
CPU max MHz:                          2001.0000
CPU min MHz:                          800.0000
BogoMIPS:                             4000.00

Versions of relevant libraries:
[pip3] hypothesis-torch==0.7.12
[pip3] mypy==1.11.2
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.0.1
[pip3] pytorch-lightning==2.4.0
[pip3] torch==2.4.1
[pip3] torchmetrics==1.4.1
[pip3] torchvision==0.19.0
[pip3] triton==3.0.0
[conda] hypothesis-torch          0.7.12                   pypi_0    pypi
[conda] numpy                     2.0.1                    pypi_0    pypi
[conda] pytorch-lightning         2.4.0                    pypi_0    pypi
[conda] torch                     2.4.1                    pypi_0    pypi
[conda] torchmetrics              1.4.1                    pypi_0    pypi
[conda] torchvision               0.19.0                   pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @kwen2501 @chauhang

**Comment by awgu** [comment]
I will be out today, but I will try to look at this beginning of next week.

**Comment by awgu** [comment]
For terminology, FSDP considers two kinds of state dicts:
- Full state dict: values are unsharded `torch.Tensor`s
- Sharded state dict: values are sharded `DTensor`s (sharded on dim-0)

FSDP1 allows saving and loading these two kinds of state dicts configured via `StateDictConfig`.

FSDP2's parameter representation directly matches the sharded state dict.
- Calling `model.state_dict()` on an FSDP2-sharded model returns a sharded state dict without any computation or communication.
- Conversely, FSDP2 only supports loading a sharded state dict.

All this means is that the conversion between full and sharded state dicts should happen outside of FSDP2. One way to do this conversion is to load the full state dict on rank 0 and iteratively broadcast and shard it on all ranks; this is to avoid loading the same full state dict in CPU RAM multiple times per host.

Here is an example of how to do this:
https://github.com/pytorch/pytorch/blob/3bc073d7280d36eec2af3b1969c165b336a8ecae/test/distributed/_composable/fsdp/test_fully_shard_init.py#L701

You may customize this logic further to support specific constraints. For example, torchtune has this logic ([code](https://github.com/pytorch/torchtune/blob/bf93806265f5ae73f51ee5a3dc256d49881319fe/torchtune/training/_distributed.py#L282)) with some special support for `NF4Tensor`s.

**Comment by ringohoffman** [comment]
It makes sense to use the sharded meta tensors of the model to inform how to distribute a loaded non-sharded state dict. But how should I distribute loaded non-sharded optimizer state tensors? The optimizer state starts out empty so I don't have any sharded meta tensors to use to inform how to shard.

**Comment by awgu** [comment]
For optimizer state dict, if you want to roll your own solution, then you could base it off the model's sharding.

Otherwise, I think distributed state dict provides utilities for this. cc: @wz337 @fegin for example API usage

**Comment by wz337** [comment]
@ringohoffman 

For both model state and optimizer state (optimizer state initialization), you can use distributed state dict. `get_optimizer_state_dict` would initialize the states for you. `set_optimizer_state_dict` would load the optim state dict into the optimizer. 

Here is a test case(https://github.com/pytorch/pytorch/blob/main/test/distributed/_composable/test_composability/test_2d_composability.py#L410) for loading a full state dict into a 2D parallelized (FSDP2+TP) model. Usage should be identical. 

For more details, you could refer to the doc here: https://pytorch.org/docs/stable/distributed.checkpoint.html#torch.distributed.checkpoint.state_dict.get_state_dict

**Comment by awgu** [comment]
Feel free to reopen if there is still an issue.

**Comment by Edenzzzz** [comment]
@awgu I'm a bit surprised to see that we still have to broadcast  before calling `distributed_tensor`, instead of having full tensor on only one rank and scatter it inside `distributed_tensor`

**Comment by Edenzzzz** [comment]
I tried loading weights on rank 0 and broadcasting to the node, but it ends up slower than all GPUs loading directly from the disk, probably due to the slow ring algo of NCCL

**Comment by ringohoffman** [comment]
> I tried loading weights on rank 0 and broadcasting to the node, but it ends up slower than all GPUs loading directly from the disk, probably due to the slow ring algo of NCCL

How big is your weight file and how many processes are you loading on? It definitely scales to be faster than reading from disk on all ranks.

**Comment by Edenzzzz** [comment]
> How big is your weight file and how many processes are you loading on? It definitely scales to be faster than reading from disk on all ranks.

14B model. I just realized this is due to RAID 1, broadcast should still be faster normally. 
https://github.com/hao-ai-lab/FastVideo/pull/470
